vm/interpreter/src/default_syscalls.rs

```diff
@@ -0,0 +1,201 @@
+                epoch: bh_2.epoch(),
+                fault_type: ConsensusFaultType::DoubleForkMining,
+            }))
+        }
+        // (b) time-offset mining fault
+        // strictly speaking no need to compare heights based on double fork mining check above,
+        // but at same height this would be a different fault.
+        else if bh_1.parents() != bh_2.parents() && bh_1.epoch() != bh_2.epoch() {
+            Ok(Some(ConsensusFault {
+                target: *bh_1.miner_address(),
+                epoch: bh_2.epoch(),
+                fault_type: ConsensusFaultType::TimeOffsetMining,
+            }))
+        }
+        // (c) parent-grinding fault
+        // Here extra is the "witness", a third block that shows the connection between A and B as
+        // A's sibling and B's parent.
+        // Specifically, since A is of lower height, it must be that B was mined omitting A from its tipset
+        else if !extra.is_empty() {
+            let bh_3: BlockHeader = from_slice(extra).map_err(|e| {
+                ActorError::new(
+                    ExitCode::ErrPlaceholder,
+                    format!("cannot decode extra {}", e.to_string()),
+                )
+            })?;
+            if bh_1.parents() != bh_3.parents()
+                && bh_1.epoch() != bh_3.epoch()
+                && contains_cid(bh_2.parents().cids(), bh_3.cid())
+                && !contains_cid(bh_2.parents().cids(), bh_1.cid())
+            {
+                Ok(Some(ConsensusFault {
+                    target: *bh_1.miner_address(),
+                    epoch: bh_2.epoch(),
+                    fault_type: ConsensusFaultType::ParentGrinding,
+                }))
+            } else {
+                Ok(None)
+            }
+        } else {
+            // (4) expensive final checks
+
+            // check blocks are properly signed by their respective miner
+            // note we do not need to check extra's: it is a parent to block b
+            // which itself is signed, so it was willingly included by the miner
+            self.verify_block_signature(&bh_1)?;
+            self.verify_block_signature(&bh_2)?;
+
+            Ok(None)
+        }
+    }
+    fn verify_block_signature(&self, bh: &BlockHeader) -> Result<(), ActorError> {
+        let state = StateTree::new_from_root(self.store, bh.state_root()).map_err(|e| {
+            ActorError::new(
+                ExitCode::ErrPlaceholder,
+                format!("cannot retrieve actor state {}", e),
+            )
+        })?;
+
+        let actor = state
+            .get_actor(bh.miner_address())
+            .map_err(|e| {
+                ActorError::new(
+                    ExitCode::ErrPlaceholder,
+                    format!("cannot retrieve actor state {}", e),
+                )
+            })?
+            .ok_or_else(|| {
+                ActorError::new(
+                    ExitCode::ErrPlaceholder,
+                    "cannot retrieve actor state".to_owned(),
```

Comment:
error message doesn't need to match, if doing something else can be more specific that it was not found and not just some error